### PR TITLE
brew-cask: move to using tap cmd directory.

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
This provides a few benefits:
- faster `brew cask` execution times as another Ruby process is not needed. Cask can instead be loaded in-process with Homebrew. This will also make it easier to use some of Homebrew's core code and ease moving code from Cask into Homebrew core.
- Users do not need to `brew upgrade` Cask any more: it's done automatically on any `brew update` or `git pull` of the Cask tap.

I appreciate this PR will probably not be merged as-is while we work out how to handle e.g. the Formula and I answer other questions.

CC @vitorgalvao @DomT4 @xu-cheng

### CODE REVIEW

**Comment by xu-cheng** [comment]
Probably want to update `brew-cask` formula to an dummy one, i.e.  bumping the version number but installing nothing.

**Comment by mikemcquaid** [comment]
> Probably want to update brew-cask formula to an dummy one, i.e. bumping the version number but installing nothing.

I was thinking the same :+1:

**Comment by jawshooah** [comment]
One question about @xu-cheng's suggestion: what about existing installations which symlink to the `brew-cask` script? Could the upgrade remove them?

**Comment by jawshooah** [comment]
Also, the failing test is due to the `cmd` directory not being listed [here](https://github.com/caskroom/homebrew-cask/blob/7adc1325314ba0bd675d7d205008540f314bb3df/test/layout_test.rb#L27).

**Comment by xu-cheng** [comment]
> what about existing installations which symlink to the brew-cask script? Could the upgrade remove them?

The symlink will be removed. The only residual file will be an empty `/usr/local/Cellar/brew-cask/version/` directory.

**Comment by mikemcquaid** [comment]
> Also, the failing test is due to the cmd directory not being listed here.

@jawshooah Yeh, will fix up the CI and do more testing.

**Comment by vitorgalvao** [comment]
I should probably mention I’ve already seen this as well. I like the idea and have nothing to add.

I’ll ask that you please always @mention @jawshooah in these situations as well, since he’s taking care of most of the actual implementation and revisions to the core.

**Comment by DomT4** [comment]
Don't have much to comment beyond to say the idea here is great :+1:.

**Comment by mikemcquaid** [comment]
Sure thing @vitorgalvao, thanks. I just mentioned you in this case as you commented on the Homebrew issue that provoked this.

**Comment by mikemcquaid** [comment]
This is green now. The only outstanding issue I noticed from the formula is how we want to handle the manpage. CC @xu-cheng for ideas.

**Comment by mikemcquaid** [comment]
Can one of the Homebrew Cask maintainers let me know what needs to be changed (except the `man` fix) to make this suitable for merging?

**Comment by jawshooah** [comment]
LGTM. Not sure how manpages are usually done for external brew commands, but we can open a separate issue for that. @vitorgalvao, any objections to merging this now?

**Comment by vitorgalvao** [comment]
No objections at all. I was also waiting on the mentioned manpage ideas, but I agree that can be a different issue.

**Comment by jawshooah** [comment]
With this merged, the version number will no longer be relevant. We'll need to bump it once to prompt users to upgrade, but after that it will have no further use. How do you want to handle that?

**Comment by vitorgalvao** [comment]
Ah, yes, hadn’t start thinking about that transition yet. Any suggestions?

**Comment by jawshooah** [comment]
`HBC_VERSION` doesn't appear to be used anywhere apart from generating the output for `brew cask --version` and `brew cask doctor`, so we may as well keep it around. Homebrew's version hasn't been updated [since 2013](https://github.com/Homebrew/homebrew/commit/cf8493d63dbed1b84690da6b3bfa211ba05efa54), but they haven't removed it.

We can get rid of [releasing.md](https://github.com/caskroom/homebrew-cask/blob/19a4da0396391a356bf70cc50e831696e616e643/doc/releasing.md) and the [bump_version](https://github.com/caskroom/homebrew-cask/blob/19a4da0396391a356bf70cc50e831696e616e643/developer/bin/bump_version), [generate_changelog](https://github.com/caskroom/homebrew-cask/blob/19a4da0396391a356bf70cc50e831696e616e643/developer/bin/generate_changelog), and [get_release_tag](https://github.com/caskroom/homebrew-cask/blob/19a4da0396391a356bf70cc50e831696e616e643/developer/bin/get_release_tag) developer scripts, as they won't be needed anymore.

Our final release notes should emphasize this change, and that there will be no more releases in the future.

**Comment by adityadalal924** [comment]
How do Cask DSL changes work if we can't bump the version anymore (something that would be required for #15643)

**Comment by jawshooah** [comment]
@adityadalal924 DSL versioning is [completely independent](https://github.com/caskroom/homebrew-cask/blob/ccaf00a189cdbe50109ba6bf294c2b72aecba784/lib/hbc/source/path_base.rb#L57) from Cask's own version.

**Comment by jawshooah** [comment]
It's important to note, however, that [we aren't currently performing any meaningful validation of the DSL version](https://github.com/caskroom/homebrew-cask/pull/13613#issuecomment-137982673). I started working on it a while ago, but haven't gotten around to finishing up.

**Comment by xu-cheng** [comment]
About manpage, a possible workaround is to auto install it whenever invoking `brew cask` command.

About version, you may want to create a develop branch for core code development.

**Comment by mikemcquaid** [comment]
> We can get rid of releasing.md and the bump_version, generate_changelog, and get_release_tag developer scripts, as they won't be needed anymore.

I'll remove these.

> About manpage, a possible workaround is to auto install it whenever invoking brew cask command.

@xu-cheng as this case seems relatively common I think I'll add an optional `man` directory to the tap root which is autosymlinked. I think that should probably block this being merged.

**Comment by mikemcquaid** [comment]
Opened a Homebrew PR for linking manpages in https://github.com/Homebrew/homebrew/pull/46795.

**Comment by mikemcquaid** [comment]
@jawshooah I think I've done everything you suggested and made sure `brew cask doctor` was clean for me.

File: lib/hbc/version.rb [hunk 1]
```diff
@@ -1 +1,13 @@
-HBC_VERSION = '0.59.0'
+HBC_VERSION = '0.59.1'
+
+module Hbc
+  def self.full_version
+    @full_version ||= begin
+      revision, commit = Hbc.default_tappath.cd do
```
Comment (jawshooah) [inline lib/hbc/version.rb]: We don't have Homebrew's `Pathname` extensions, so this line breaks the build.

File: lib/hbc/version.rb [hunk 2]
```diff
@@ -1 +1,13 @@
...
+        [`git rev-parse --short=4 --verify -q HEAD 2>/dev/null`.chuzzle,
```
Comment (jawshooah) [inline lib/hbc/version.rb]: Same here; we haven't extended `String`. `chomp` should work fine, though.

File: lib/hbc/version.rb (hunk 1 above)
Comment (mikemcquaid) [inline lib/hbc/version.rb]: Heh, worked for me locally because this is loaded in the same process as Homebrew. I'll fix it up regardless.

**Comment by mikemcquaid** [comment]
@jawshooah Comments addressed (and I only just "got" your username)

**Comment by jawshooah** [comment]
Nice, just waiting on Travis to give me the OK.

And yeah, I've noticed a lot of people read it as "Jaws, Hoo-ah!" Like I'm a Marine who's just really enthusiastic about Steven Spielberg films.

**Comment by jawshooah** [comment]
Ah, `version_test.rb` will [need to be updated](https://travis-ci.org/caskroom/homebrew-cask/jobs/95611071#L2472).

**Comment by mikemcquaid** [comment]
Tests are :green_heart:

**Comment by mikemcquaid** [comment]
Note: this is blocked on https://github.com/Homebrew/homebrew/pull/46795 and when this is merged you should do an immediate tag and final release of `brew-cask.rb`.

**Comment by jawshooah** [comment]
Not that it matters much, but I think this change warrants a bump to `0.60.0` rather than `0.59.1`. It's certainly more than a patch, and 60 is a nice, round number.

**Comment by mikemcquaid** [comment]
@jawshooah Updated to 0.60.0 and https://github.com/Homebrew/homebrew/pull/46795 was merged. When you're :ok:: merge, tag and shout and I'll be able to remove some code from Homebrew too. Thanks for the review here!

**Comment by jawshooah** [comment]
Great! So will users have to `brew update && brew uninstall brew-cask && brew tap --repair caskroom/cask`?

**Comment by mikemcquaid** [comment]
`brew update` will take care of the linking but `brew tap --repair` will link it without an `update`. They technically won't _need_ to `brew uninstall` but that package will be empty/pointless so should do so when they get a chance (or `cask` could maybe even do that for them).

**Comment by jawshooah** [comment]
Does Homebrew display `caveats` on `brew upgrade`? If not, users won't see the message informing them that the `brew-cask` package is now obsolete.

We could add something like the following to `cmd/brew-cask.rb`:

```rb
`command -v brew-cask`
puts <<-EOS.undent if $?.success?
  You can `brew uninstall brew-cask`, as `brew tap caskroom/cask` is now all that's
  needed to install Homebrew Cask and keep it up to date.
EOS
```

Could you add that before we merge this?

**Comment by mikemcquaid** [comment]
Caveats are showed on `upgrade`, yep.

**Comment by mikemcquaid** [comment]
> Could you take care of that before we merge this?

Happy to still do that, though, if you'd like.

**Comment by mikemcquaid** [comment]
> Happy to still do that, though, if you'd like.

(although it feels like `cask doctor` might be a good fit for that?

**Comment by jawshooah** [comment]
> although it feels like cask doctor might be a good fit for that?

Ah, good point. You can cherry-pick my patch for `lib/cli/doctor.rb` [here](https://github.com/jawshooah/homebrew-cask/commit/7a2bb5a0e330a066fa986f5301db6010f93bc8cf).

`developer/bin/develop_brew_cask` also needs some fixing up to reflect this change. Patch is [here](https://github.com/jawshooah/homebrew-cask/commit/7db3e8eb1b5833433e4a7d4fe40439320901ae64).

**Comment by mikemcquaid** [comment]
@jawshooah Pushed.

**Comment by jawshooah** [comment]
@vitorgalvao, I'm giving you the final word on this. Ready to merge?

**Comment by vitorgalvao** [comment]
Please, you should be the one to do the honours of merging.

Thank you @jawshooah and @mikemcquaid for all the work and coordination on this.

**Comment by mikemcquaid** [comment]
@vitorgalvao Thanks for the kind words, great project and @jawshooah thanks for the merge!

**Comment by DomT4** [comment]
Seeing some issues locally:

```
Updated 1 tap (caskroom/cask).
Error: Could not link caskroom/cask manpages to:
  /usr/local/share/man/man1/brew-cask.1

Please delete these files and run `brew tap --repair`.
No changes to formulae.
==> Upgrading 1 outdated package, with result:
caskroom/cask/brew-cask 0.60.0
==> Upgrading caskroom/cask/brew-cask
rm /usr/local/bin/brew-cask
rm /usr/local/share/man/man1/brew-cask.1
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20151209-9440-rnz8z0.sb nice /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Library/Homebrew -- /usr/local/Library/Homebrew/build.rb /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb --verbose
==> Cloning https://github.com/caskroom/homebrew-cask.git
git clone --depth 1 --branch v0.60.0 https://github.com/caskroom/homebrew-cask.git /usr/local/var/homebrew/cache/brew-cask--git
Cloning into '/usr/local/var/homebrew/cache/brew-cask--git'...
warning: Could not find remote branch v0.60.0 to clone.
fatal: Remote branch v0.60.0 not found in upstream origin
Error: Failed to download resource "brew-cask"
Failure while executing: git clone --depth 1 --branch v0.60.0 https://github.com/caskroom/homebrew-cask.git /usr/local/var/homebrew/cache/brew-cask--git
ln -s ../Cellar/brew-cask/0.59.0/bin/brew-cask brew-cask
ln -s ../../../Cellar/brew-cask/0.59.0/share/man/man1/brew-cask.1 brew-cask.1
```

**Comment by DomT4** [comment]
Might be resolved by https://github.com/caskroom/homebrew-cask/commit/f2c2f85b977de30ba1aa65bd6a92e40925b21a57. Double checking now.

**Comment by DomT4** [comment]
:+1:. Got the manpage link failure still but otherwise all issues resolved. Probably fairly unlikely anyone who wasn't watching this issue tried to update in the last 7 minutes, heh.

**Comment by fanquake** [comment]
:+1:

**Comment by swrobel** [comment]
@mikemcquaid this is awesome! :clap: Merry christmas to us all! :santa:

**Comment by naffan2014** [comment]
不错的想法！ 虽然我不怎么懂吧

**Comment by vitorgalvao** [comment]
@naffan2014 English only, please. Google Translate returns “Good idea! Although I do not understand it”.

In essence, `brew update` will also now upgrade homebrew-cask automatically.

**Comment by naffan2014** [comment]
lol. google is great . it translate  what i mean exactly .haha

**Comment by hovancik** [comment]
so i tried this today, but I am getting error: 

```
$ brew uninstall --force brew-cask; brew update
Already up-to-date.
$ brew cask install Caskroom/cask/flash-player
Error: Unknown command: cask
$ brew cask install Caskroom/cask/flash-player --verbose
Error: Unknown command: cask
Error: Kernel.exit
$ brew update; brew cleanup; brew cask cleanup
Already up-to-date.
Warning: Skipping (old) keg-only: /usr/local/Cellar/openssl/1.0.2d_1
Error: Unknown command: cask
$ brew cask doctor
Error: Unknown command: cask
```

**Comment by adityadalal924** [comment]
@hovancik and others: Please make a new issue so we can avoid adding unrelated comments to this (merged) PR.

**Pull request merged**